File: CSharp/Gravical.Poker.Core.Tests/TableFinalTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Diagnostics.CodeAnalysis;


namespace Gravical.Poker.Core.Tests
{
    [TestClass]
    [ExcludeFromCodeCoverage]
    public class TableFinalTests : TestBase
    {
        private TableFinal Simple = new TableFinal(
            new Card(Face.Two, Suit.Heart),
            new Card(Face.Three, Suit.Club),
            new Card(Face.Four, Suit.Diamond),
            new Card(Face.Five, Suit.Spade),
            new Card(Face.Six, Suit.Heart));

        [TestMethod]
        public void Ctor_Normal_ShouldSetCorrectCards()
        {
            Assert.AreEqual(new Card(Face.Two, Suit.Heart), Simple.First);
            Assert.AreEqual(new Card(Face.Three, Suit.Club), Simple.Second);
            Assert.AreEqual(new Card(Face.Four, Suit.Diamond), Simple.Third);
            Assert.AreEqual(new Card(Face.Five, Suit.Spade), Simple.Turn);
            Assert.AreEqual(new Card(Face.Six, Suit.Heart), Simple.River);
        }

        [TestMethod]
        public void ToArray_Normal_ShouldReturnInOrder()
        {
            var actual = Simple.ToArray();
            Assert.AreEqual(5, actual.Length);
            Assert.AreEqual(new Card(Face.Two, Suit.Heart), actual[0]);
            Assert.AreEqual(new Card(Face.Three, Suit.Club), actual[1]);
            Assert.AreEqual(new Card(Face.Four, Suit.Diamond), actual[2]);
            Assert.AreEqual(new Card(Face.Five, Suit.Spade), actual[3]);
            Assert.AreEqual(new Card(Face.Six, Suit.Heart), actual[4]);
        }

        [TestMethod]
        public void GetAllCards_Normal_ShouldReturnSevenCards()
        {
            var arranged = new Pocket(new Deck());
            var actual = Simple.GetAllCards(arranged);
            Assert.AreEqual(7, actual.Length);
            Assert.AreEqual(new Card(Face.Two, Suit.Heart), actual[0]);
            Assert.AreEqual(new Card(Face.Two, Suit.Heart), actual[0]);
            Assert.AreEqual(new Card(Face.Three, Suit.Club), actual[1]);
            Assert.AreEqual(new Card(Face.Four, Suit.Diamond), actual[2]);
            Assert.AreEqual(new Card(Face.Five, Suit.Spade), actual[3]);
            Assert.AreEqual(new Card(Face.Six, Suit.Heart), actual[4]);
            Assert.AreEqual(arranged.CardA, actual[5]);
            Assert.AreEqual(arranged.CardB, actual[6]);
        }

        [TestMethod]
        public void Contains_DoesNot_ShouldReturnFalse()
        {
            Assert.IsFalse(Simple.Contains(new Card(Face.Ace, Suit.Heart)));
        }

        [TestMethod]
        public void Contains_MatchFirst_ShouldReturnTrue()
        {
            Assert.IsTrue(Simple.Contains(new Card(Face.Two, Suit.Heart)));
        }

        [TestMethod]
        public void Contains_MatchSecond_ShouldReturnTrue()
        {
            Assert.IsTrue(Simple.Contains(new Card(Face.Three, Suit.Club)));
        }

        [TestMethod]
        public void Contains_MatchThird_ShouldReturnTrue()
        {
            Assert.IsTrue(Simple.Contains(new Card(Face.Four, Suit.Diamond)));
        }

        [TestMethod]
        public void Contains_MatchTurn_ShouldReturnTrue()
        {
            Assert.IsTrue(Simple.Contains(new Card(Face.Five, Suit.Spade)));
        }

        [TestMethod]
        public void Contains_MatchRiver_ShouldReturnTrue()
        {
            Assert.IsTrue(Simple.Contains(new Card(Face.Six, Suit.Heart)));
        }

        [TestMethod]
        public void ToString_Normal_ShouldReturnKnownString()
        {
            Assert.AreEqual("2h 3c 4d 5s 6h", Simple.ToString());
        }
    }
}

File: JavaScript/test/TableFinalTests.mjs
import { strict as assert } from 'assert';
import { Tests } from './TestFixture.mjs';
import { Enums } from '../src/Enums.mjs';
import { Card } from '../src/Card.mjs';
import { Pocket } from '../src/Pocket.mjs';
import { Deck } from '../src/Deck.mjs';
import { TableFinal } from '../src/TableFinal.mjs';

function simple() {
	return TableFinal.make(
		Card.parse('2h'),
		Card.parse('3c'),
		Card.parse('4d'),
		Card.parse('5s'),
		Card.parse('6h'));
}

Tests.add('TableFinal.constructor() should set all cards null', () => {
	let actual = new TableFinal();
	assert.ok(!actual.first);
	assert.ok(!actual.second);
	assert.ok(!actual.third);
	assert.ok(!actual.turn);
	assert.ok(!actual.river);
});

Tests.add('TableFinal.make() should set correct cards', () => {
	assert.equal(Card.parse('2h').uint8(), simple().first.uint8());
	assert.equal(Card.parse('3c').uint8(), simple().second.uint8());
	assert.equal(Card.parse('4d').uint8(), simple().third.uint8());
	assert.equal(Card.parse('5s').uint8(), simple().turn.uint8());
	assert.equal(Card.parse('6h').uint8(), simple().river.uint8());
});

Tests.add('TableFinal.toArray() null should return all null', () => {
	let actual = new TableFinal().toArray();
	assert.equal(5, actual.length);
	assert.ok(!actual[0]);
	assert.ok(!actual[1]);
	assert.ok(!actual[2]);
	assert.ok(!actual[3]);
	assert.ok(!actual[4]);
});

Tests.add('TableFinal.toArray() should return in order', () => {
	let actual = simple().toArray();
	assert.equal(5, actual.length);
	assert.equal(Card.parse('2h').uint8(), actual[0].uint8());
	assert.equal(Card.parse('3c').uint8(), actual[1].uint8());
	assert.equal(Card.parse('4d').uint8(), actual[2].uint8());
	assert.equal(Card.parse('5s').uint8(), actual[3].uint8());
	assert.equal(Card.parse('6h').uint8(), actual[4].uint8());
});

Tests.add('TableFinal.getAllCards() invalid pocket should throw', () => {
	assert.throws(() => simple().getAllCards(null), /^Pocket isn't valid$/);
	assert.throws(() => simple().getAllCards(new Object()), /^Pocket isn't valid$/);
	assert.throws(() => simple().getAllCards('example'), /^Pocket isn't valid$/);
	assert.throws(() => simple().getAllCards(123), /^Pocket isn't valid$/);
});

Tests.add('TableFinal.getAllCards() should return seven cards', () => {
	let arranged = new Pocket(new Deck());
	let actual = simple().getAllCards(arranged);
	assert.equal(7, actual.length);
	assert.equal(Card.parse('2h').uint8(), actual[0].uint8());
	assert.equal(Card.parse('2h').uint8(), actual[0].uint8());
	assert.equal(Card.parse('3c').uint8(), actual[1].uint8());
	assert.equal(Card.parse('4d').uint8(), actual[2].uint8());
	assert.equal(Card.parse('5s').uint8(), actual[3].uint8());
	assert.equal(Card.parse('6h').uint8(), actual[4].uint8());
	assert.equal(arranged.a.uint8(), actual[5].uint8());
	assert.equal(arranged.b.uint8(), actual[6].uint8());
});

Tests.add('TableFinal.contains() not a card should return false', () => {
	assert.ok(!simple().contains(null));
	assert.ok(!simple().contains(new Object()));
	assert.ok(!simple().contains('example'));
	assert.ok(!simple().contains(123));
});

Tests.add('TableFinal.contains() does not should return false', () => {
	assert.ok(!simple().contains(Card.parse('Ah')));
});

Tests.add('TableFinal.contains() match first should return true', () => {
	assert.ok(simple().contains(Card.parse('2h')));
});

Tests.add('TableFinal.contains() match second should return true', () => {
	assert.ok(simple().contains(Card.parse('3c')));
});

Tests.add('TableFinal.contains() match third should return true', () => {
	assert.ok(simple().contains(Card.parse('4d')));
});

Tests.add('TableFinal.contains() match turn should return true', () => {
	assert.ok(simple().contains(Card.parse('5s')));
});

Tests.add('TableFinal.contains() match river should return true', () => {
	assert.ok(simple().contains(Card.parse('6h')));
});

Tests.add('TableFinal.toString() null should return nulls', () => {
	assert.equal("null null null null null", new TableFinal().toString());
});

Tests.add('TableFinal.toString() should return correct string', () => {
	assert.equal("2h 3c 4d 5s 6h", simple().toString());
});
